File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/CS/MessageInspectors/TransactionFlowInspector.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.ServiceModel;
using System.ServiceModel.Dispatcher;
using System.Transactions;

namespace Microsoft.ServiceModel.Samples
{
    class TransactionFlowInspector : IClientMessageInspector
    {
        public void AfterReceiveReply(ref System.ServiceModel.Channels.Message reply, object correlationState)
        {
        }

        public object BeforeSendRequest(ref System.ServiceModel.Channels.Message request, System.ServiceModel.IClientChannel channel)
        {
            // obtain the tx propagation token
            byte[] propToken = null;           
            if (Transaction.Current != null && IsTxFlowRequiredForThisOperation(request.Headers.Action))
            {
                try
                {
                    propToken = TransactionInterop.GetTransmitterPropagationToken(Transaction.Current);
                }
                catch (TransactionException e)
                {
                    throw new CommunicationException("TransactionInterop.GetTransmitterPropagationToken failed.", e);
                }
            }

            // set the propToken on the message in a TransactionFlowProperty
            TransactionFlowProperty.Set(propToken, request);

            return null;            
        }

        static bool IsTxFlowRequiredForThisOperation(String action)
        {
            // In general, this should contain logic to identify which operations (actions) require transaction flow.
            // Here we just flow transactions for all actions.
            return true;
        }
    }
}

File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/MessageInspectors/TransactionFlowInspector.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.ServiceModel
Imports System.ServiceModel.Dispatcher
Imports System.Transactions

Namespace Microsoft.ServiceModel.Samples
	Friend Class TransactionFlowInspector
		Implements IClientMessageInspector
        Public Sub AfterReceiveReply(ByRef reply As System.ServiceModel.Channels.Message,
                                     ByVal correlationState As Object) Implements IClientMessageInspector.AfterReceiveReply
        End Sub

        Public Function BeforeSendRequest(ByRef request As System.ServiceModel.Channels.Message,
                                          ByVal channel As System.ServiceModel.IClientChannel) As Object Implements IClientMessageInspector.BeforeSendRequest
            ' obtain the tx propagation token
            Dim propToken() As Byte = Nothing
            If Transaction.Current IsNot Nothing AndAlso IsTxFlowRequiredForThisOperation(request.Headers.Action) Then
                Try
                    propToken = TransactionInterop.GetTransmitterPropagationToken(Transaction.Current)
                Catch e As TransactionException
                    Throw New CommunicationException("TransactionInterop.GetTransmitterPropagationToken failed.", e)
                End Try
            End If

            ' set the propToken on the message in a TransactionFlowProperty
            TransactionFlowProperty.Set(propToken, request)

            Return Nothing
        End Function

		Private Shared Function IsTxFlowRequiredForThisOperation(ByVal action As String) As Boolean
			' In general, this should contain logic to identify which operations (actions) require transaction flow.
			' Here we just flow transactions for all actions.
			Return True
		End Function
	End Class
End Namespace
